webapp/components/ContentMenu/ContentMenu.spec.js

```diff
@@ -0,0 +1,463 @@
+        const wrapper = openContentMenu({
+          isOwner: false,
+          resourceType: 'contribution',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+            pinnedBy: null,
+          },
+        })
+        wrapper
+          .findAll('.ds-menu-item')
+          .filter(item => item.text() === 'post.menu.pin')
+          .at(0)
+          .trigger('click')
+        expect(wrapper.emitted('pinPost')).toEqual([
+          [
+            {
+              id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+              pinnedBy: null,
+            },
+          ],
+        ])
+      })
+
+      it('unpin pinned post', () => {
+        const wrapper = openContentMenu({
+          isOwner: false,
+          resourceType: 'contribution',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+            pinnedBy: 'someone',
+          },
+        })
+        wrapper
+          .findAll('.ds-menu-item')
+          .filter(item => item.text() === 'post.menu.unpin')
+          .at(0)
+          .trigger('click')
+        expect(wrapper.emitted('unpinPost')).toEqual([
+          [
+            {
+              id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+              pinnedBy: 'someone',
+            },
+          ],
+        ])
+      })
+    })
+
+    describe('owner of comment can', () => {
+      let wrapper
+      beforeEach(() => {
+        wrapper = openContentMenu({
+          isOwner: true,
+          resourceType: 'comment',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+          },
+        })
+        openModalSpy = jest.spyOn(wrapper.vm, 'openModal')
+      })
+      it('edit the comment', () => {
+        wrapper
+          .findAll('.ds-menu-item')
+          .filter(item => item.text() === 'comment.menu.edit')
+          .at(0)
+          .trigger('click')
+        expect(wrapper.emitted('showEditCommentMenu')).toEqual([[true]])
+      })
+      it('delete the comment', () => {
+        wrapper
+          .findAll('.ds-menu-item')
+          .filter(item => item.text() === 'comment.menu.delete')
+          .at(0)
+          .trigger('click')
+        expect(openModalSpy).toHaveBeenCalledWith('delete')
+      })
+    })
+
+    describe('reporting', () => {
+      it('a post of another user is possible', () => {
+        getters['auth/isAdmin'] = () => false
+        getters['auth/isModerator'] = () => false
+        const wrapper = openContentMenu({
+          isOwner: false,
+          resourceType: 'contribution',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+          },
+        })
+        openModalSpy = jest.spyOn(wrapper.vm, 'openModal')
+        wrapper
+          .findAll('.ds-menu-item')
+          .filter(item => item.text() === 'report.contribution.title')
+          .at(0)
+          .trigger('click')
+        expect(openModalSpy).toHaveBeenCalledWith('report')
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-11-11T22:00:38Z" title="Monday, November 11th 2019, 11:00:38 pm +01:00">Nov 11, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,543 @@
+import { config, mount, createLocalVue } from '@vue/test-utils'
+import Vuex from 'vuex'
+import VTooltip from 'v-tooltip'
+import Styleguide from '@human-connection/styleguide'
+import ContentMenu from './ContentMenu.vue'
+
+const localVue = createLocalVue()
+
+localVue.use(Styleguide)
+localVue.use(VTooltip)
+localVue.use(Vuex)
+
+config.stubs['router-link'] = '<span><slot /></span>'
+
+let getters, mutations, mocks, menuToggle, openModalSpy
+
+describe('ContentMenu.vue', () => {
+  beforeEach(() => {
+    mocks = {
+      $t: jest.fn(str => str),
+      $i18n: {
+        locale: () => 'en',
+      },
+      $router: {
+        resolve: jest.fn(obj => {
+          obj.href = '/post/edit/d23a4265-f5f7-4e17-9f86-85f714b4b9f8'
+          return obj
+        }),
+        push: jest.fn(),
+      },
+    }
+  })
+
+  describe('mount', () => {
+    mutations = {
+      'modal/SET_OPEN': () => jest.fn(),
+    }
+    getters = {
+      'auth/isModerator': () => false,
+      'auth/isAdmin': () => false,
+    }
+
+    const openContentMenu = (values = {}) => {
+      const store = new Vuex.Store({ mutations, getters })
+      const wrapper = mount(ContentMenu, {
+        propsData: {
+          ...values,
+        },
+        mocks,
+        store,
+        localVue,
+      })
+      menuToggle = wrapper.find('.content-menu-trigger')
+      menuToggle.trigger('click')
+      return wrapper
+    }
+
+    describe('contribution', () => {
+      it('owner can edit and delete contribution', () => {
+        const wrapper = openContentMenu({
+          isOwner: true,
+          resourceType: 'contribution',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+          },
+        })
+        let items = wrapper.findAll('.ds-menu-item')
+        openModalSpy = jest.spyOn(wrapper.vm, 'openModal')
+        expect(items).toHaveLength(2)
+        expect(items.at(0).text()).toBe('post.menu.edit')
+        expect(
+          items
+            .at(0)
+            .find('span.ds-menu-item-link')
+            .attributes('to'),
+        ).toBe('/post/edit/d23a4265-f5f7-4e17-9f86-85f714b4b9f8')
+        expect(items.at(1).text()).toBe('post.menu.delete')
+        expect(
+          items
+            .at(1)
+            .find('span.ds-menu-item-link')
+            .attributes('to'),
+        ).toBe('/post-menu-delete')
+        items.at(1).trigger('click')
+        expect(openModalSpy).toHaveBeenCalledWith('delete')
+      })
+
+      it('admin can pin unpinned post', () => {
+        getters['auth/isAdmin'] = () => true
+        const wrapper = openContentMenu({
+          isOwner: false,
+          resourceType: 'contribution',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+            pinnedBy: null,
+          },
+        })
+        let items = wrapper.findAll('.ds-menu-item')
+        expect(items.at(0).text()).toBe('post.menu.pin')
+        expect(
+          items
+            .at(0)
+            .find('span.ds-menu-item-link')
+            .attributes('to'),
+        ).toBe('/post-menu-pin')
+        items.at(0).trigger('click')
+        expect(wrapper.emitted('pinPost')).toEqual([
+          [
+            {
+              id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+              pinnedBy: null,
+            },
+          ],
+        ])
+      })
+
+      it('admin can unpin pinned post', () => {
+        const wrapper = openContentMenu({
+          isOwner: false,
+          resourceType: 'contribution',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+            pinnedBy: 'someone',
+          },
+        })
+        let items = wrapper.findAll('.ds-menu-item')
+        // for (var i = 0; i < items.length; i++) { console.log(items.at(i).html()) }
+        expect(items.at(0).text()).toBe('post.menu.unpin')
+        expect(
+          items
+            .at(0)
+            .find('span.ds-menu-item-link')
+            .attributes('to'),
+        ).toBe('/post-menu-unpin')
+        items.at(0).trigger('click')
+        expect(wrapper.emitted('unpinPost')).toEqual([
+          [
+            {
+              id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+              pinnedBy: 'someone',
+            },
+          ],
+        ])
+      })
+    })
+
+    describe('comment', () => {
+      it('owner can edit and delete comment', () => {
+        const wrapper = openContentMenu({
+          isOwner: true,
+          resourceType: 'comment',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+          },
+        })
+        openModalSpy = jest.spyOn(wrapper.vm, 'openModal')
+        let items = wrapper.findAll('.ds-menu-item')
+        expect(items).toHaveLength(2)
+        expect(items.at(0).text()).toBe('comment.menu.edit')
+        expect(
+          items
+            .at(0)
+            .find('span.ds-menu-item-link')
+            .attributes('to'),
+        ).toBe('/comment-menu-edit')
+        items.at(0).trigger('click')
+        expect(wrapper.emitted('showEditCommentMenu')).toEqual([[true]])
+        expect(items.at(1).text()).toBe('comment.menu.delete')
+        expect(
+          items
+            .at(1)
+            .find('span.ds-menu-item-link')
+            .attributes('to'),
+        ).toBe('/comment-menu-delete')
+        items.at(1).trigger('click')
+        expect(openModalSpy).toHaveBeenCalledWith('delete')
+      })
+    })
+
+    describe('report', () => {
+      it('anyone who is not owner can report post', () => {
+        getters['auth/isAdmin'] = () => false
+        getters['auth/isModerator'] = () => false
+        const wrapper = openContentMenu({
+          isOwner: false,
+          resourceType: 'contribution',
+          resource: {
+            id: 'd23a4265-f5f7-4e17-9f86-85f714b4b9f8',
+          },
+        })
+        openModalSpy = jest.spyOn(wrapper.vm, 'openModal')
+        let items = wrapper.findAll('.ds-menu-item')
+        expect(items).toHaveLength(1)
+        expect(items.at(0).text()).toBe('report.contribution.title')
+        expect(
+          items
+            .at(0)
+            .find('span.ds-menu-item-link')
+            .attributes('to'),
+        ).toBe('/report-contribution-title')
+        items.at(0).trigger('click')
+        expect(openModalSpy).toHaveBeenCalledWith('report')
```


instead of setting up the `method` openModal as a spy, can we set up `$store.commit` as a mock, and check that is called with specific data?

Reply:
> [<img alt="Mogge" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Mogge) **Authored by [Mogge](https://github.com/Mogge)**
_<time datetime="2019-11-13T04:41:39Z" title="Wednesday, November 13th 2019, 5:41:39 am +01:00">Nov 13, 2019</time>_

---


I tried, but it is getting very complicated because y also have to mount the Modal component. I think we have to rely on the tests of the Modal component that the commits are called correctly (which as far as I can see is the case). I ensure in the tests of ContentMenu that the parameters are passed to Modal correctly.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-11-27T17:02:47Z" title="Wednesday, November 27th 2019, 6:02:47 pm +01:00">Nov 27, 2019</time>_

---


I can see that, and as we discussed today, you are testing it up to one point, and there are other tests that take it further!!